### DISCUSSION TITLE
Looking for help/guidance on writing my own i2c lib

### DISCUSSION THREAD

**Discussion by harvybob** [arendst/Tasmota#20853] [Q&A]
Hi All,

I have an Athom 8 way relay [board](https://www.athom.tech/blank-1/8ch-inching-self-lock-relay) running tasmota, which I want to use to control a further [byvac](http://www.byvac.co.uk/index.php/BV4627) relay board -  over i2c. 

I've connected it up using the standard i2c driver and found no devices, and I appreciate that very few people will have the ByVac boards so looking to write my own driver or if I can use an existing Arduino code [here](http://www.byvac.com/downloads/BV4627/BV4627_a.zip) then a bit of guidance on what I need to do to pull this in to build my own image with it included?

I have tried a few additional drivers via custom images based on the base addresses but none of those seem to have been successful in identifying the i2c device.
(This was done using the Gitpod process listed [here](https://tasmota.github.io/docs/Compile-your-build/#compiling-tools)

The base address for the byvac boards is 0x32 but they can be changed with some work

**Comment by barbudor** [comment]
If this board is using a custom I2C protocol and not a standard IO expander like PCF8574 or MCP, then you have to write your own Tasmota driver to get it supported

**Reply by harvybob** [reply]
Any tips/links on what the driver should look like? I'm hoping I can modify the one in the Arduino code to work, but not sure where to start to see an existing example, and then how to pull it together into a new "build"

**Reply by barbudor** [reply]
Like there are already 80+ I2C drivers in Tasmota that you can have a loot at.
The most related would be the IO expanders ones: [PCF8574](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_28_pcf8574_v2.ino) and [MCP23XXX](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_67_mcp23xxx.ino) but beware there a not the simplest ones as they are meant to provide inputs and outputs integrated with Tasmota's switches and relays.

It would have been easier to either chose a relay board based on a supported GPIO extender, or a main board based on ESP32 to use the Berry language
